libavformat/rtcenc.c

```diff
@@ -0,0 +1,1689 @@
+            ret = AVERROR(EIO);
+            goto end;
+        }
+    }
+
+    rtc->sdp_answer = av_strdup(tmp);
+    if (!rtc->sdp_answer) {
+        ret = AVERROR(ENOMEM);
+        goto end;
+    }
+    av_log(s, AV_LOG_VERBOSE, "WHIP: Got answer: %s\n", rtc->sdp_answer);
+
+end:
+    ffurl_closep(&whip_uc);
+    av_free(tmp);
+    return ret;
+}
+
+/**
+ * Parse the ice ufrag, pwd and candidates from the answer.
+ *
+ * @return 0 if OK, AVERROR_xxx on error
+ */
+static int parse_answer(AVFormatContext *s)
+{
+    int ret = 0;
+    AVIOContext *pb;
+    char line[MAX_URL_SIZE];
+    const char *ptr;
+    int i;
+    RTCContext *rtc = s->priv_data;
+
+    pb = avio_alloc_context(rtc->sdp_answer, strlen(rtc->sdp_answer), 0, NULL, NULL, NULL, NULL);
+    if (!pb)
+        return AVERROR(ENOMEM);
+
+    for (i = 0; !avio_feof(pb); i++) {
+        ff_get_chomp_line(pb, line, sizeof(line));
+        if (av_strstart(line, "a=ice-ufrag:", &ptr) && !rtc->ice_ufrag_remote) {
+            rtc->ice_ufrag_remote = av_strdup(ptr);
+            if (!rtc->ice_ufrag_remote) {
+                ret = AVERROR(ENOMEM);
+                goto end;
+            }
+        } else if (av_strstart(line, "a=ice-pwd:", &ptr) && !rtc->ice_pwd_remote) {
+            rtc->ice_pwd_remote = av_strdup(ptr);
+            if (!rtc->ice_pwd_remote) {
+                ret = AVERROR(ENOMEM);
+                goto end;
+            }
+        } else if (av_strstart(line, "a=candidate:", &ptr) && !rtc->ice_protocol) {
+            ptr = av_stristr(ptr, "udp");
+            if (ptr && av_stristr(ptr, "host")) {
+                char protocol[17], host[129];
+                int priority, port;
+                ret = sscanf(ptr, "%16s %d %128s %d typ host", protocol, &priority, host, &port);
+                if (ret != 4) {
+                    av_log(s, AV_LOG_ERROR, "Failed %d to parse line %d %s from %s\n",
+                        ret, i, line, rtc->sdp_answer);
+                    ret = AVERROR(EIO);
+                    goto end;
+                }
+
+                if (av_strcasecmp(protocol, "udp")) {
+                    av_log(s, AV_LOG_ERROR, "Protocol %s is not supported by RTC, choose udp, line %d %s of %s\n",
+                        protocol, i, line, rtc->sdp_answer);
+                    ret = AVERROR(EIO);
+                    goto end;
+                }
+
+                rtc->ice_protocol = av_strdup(protocol);
+                rtc->ice_host = av_strdup(host);
+                rtc->ice_port = port;
+                if (!rtc->ice_protocol || !rtc->ice_host) {
+                    ret = AVERROR(ENOMEM);
+                    goto end;
+                }
+            }
+        }
+    }
+
+    av_log(s, AV_LOG_INFO, "WHIP: SDP offer=%luB, answer=%luB, ufrag=%s, pwd=%luB, transport=%s://%s:%d\n",
+        strlen(rtc->sdp_offer), strlen(rtc->sdp_answer), rtc->ice_ufrag_remote, strlen(rtc->ice_pwd_remote),
+        rtc->ice_protocol, rtc->ice_host, rtc->ice_port);
```

Comment:
[rtc @ 0x6170000cda80] WHIP: Init ice_arq_max=5, ice_arq_timeout=30, dtls_arq_max=5, dtls_arq_timeout=50 pkt_size=1500
AddressSanitizer:DEADLYSIGNAL
=================================================================
==29537==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x7ff819e16572 bp 0x7ff7b65e7b20 sp 0x7ff7b65e7b20 T0)
==29537==The signal is caused by a READ memory access.
==29537==Hint: address points to the zero page.
    #0 0x7ff819e16572 in _platform_strlen+0x12 (libsystem_platform.dylib:x86_64+0x1572) (BuildId: f314b62b98f43a7c82968739f8b6855a240000001000000000010d0000010d00)
    #1 0x10f69e4c7 in wrap_strlen+0x37 (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x194c7) (BuildId: 756bb7515781379f84412f22c4274ffd2400000010000000000a0a0000030d00)
    #2 0x109fb95bc in parse_answer rtcenc.c:677
    #3 0x10b8a0193 in rtc_init rtcenc.c:1559
    #4 0x109f64522 in init_muxer mux.c:350
    #5 0x109f63ac3 in avformat_init_output mux.c:438
    #6 0x109f64857 in avformat_write_header mux.c:462
    #7 0x10992e30f in mux_check_init ffmpeg_mux.c:590
    #8 0x10992ea61 in of_stream_init ffmpeg_mux.c:670
    #9 0x10992390e in enc_open ffmpeg_enc.c:421
    #10 0x109924876 in enc_frame ffmpeg_enc.c:1109
    #11 0x109928bb1 in reap_filters ffmpeg_filter.c:1449
    #12 0x1099526de in transcode_step ffmpeg.c:2109
    #13 0x109950f74 in transcode ffmpeg.c:2145
    #14 0x109950ab2 in main ffmpeg.c:2260
    #15 0x7ff819abb30f in start+0x97f (dyld:x86_64+0xfffffffffff7230f) (BuildId: bb7a09708c623dcea7a25cec9c501f1132000000200000000100000000010d00)

==29537==Register values:
rax = 0x000000011010827c  rbx = 0x000000010cb2d500  rcx = 0x0000000000000000  rdx = 0x0000000000000000
rdi = 0x0000000000000000  rsi = 0x00001c1c00009bf0  rbp = 0x00007ff7b65e7b20  rsp = 0x00007ff7b65e7b20
 r8 = 0x00001c1c00009bf7   r9 = 0x0000000000000004  r10 = 0x00000c1c00009be4  r11 = 0x00001c1c00009bf7
r12 = 0x000000010cb2eda0  r13 = 0x0000000000000000  r14 = 0x000000010cb2ed80  r15 = 0x0000000109950940
AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV (libsystem_platform.dylib:x86_64+0x1572) (BuildId: f314b62b98f43a7c82968739f8b6855a240000001000000000010d0000010d00) in _platform_strlen+0x12
==29537==ABORTING
Abort trap: 6

Reply:
Configure FFmpeg with Asan support:

```bash
./configure \
  --extra-cflags="-fsanitize=address -fno-omit-frame-pointer" \
  --extra-ldflags="-fsanitize=address"
```